File: Ejercicio_2.1/Solución.py
import math as mt
import matplotlib.pyplot as plt
import numpy as np

def f(x,g):
    a=mt.cosh(x)
    b=mt.cos(x)
    return a+b-g

def sign(a,b):
    if(a*b<0):
        return True
    elif(a*b>0):
        return False
    else:
        return False
    
def search(f,a,g):
    ya=f(a,g)
    bs=a
    bi=a
    limit=1e3
    k=1
    while(k<=limit):
        bi-=0.01
        bs+=0.01
        if(sign(ya,f(bs,g))):
            return bs
        elif(sign(ya,f(bi,g))):
            return bi
        k+=1
    print('Esta Función no tiene un cambio de signo')
    return None

def bisec(f,a,b,g):
    xi=a
    xf=b
    xm=(xi+xf)/2
    h=1e-10
    while(abs(f(xm,g))>=h):
        if(sign(f(xi,g),f(xm,g))):
            xf=xm
            xm=(xi+xf)/2
        elif(sign(f(xm,g),f(xf,g))):
            xi=xm
            xm=(xi+xf)/2
    return xm

for g in range(1,4):
    x=np.random.rand(1000)*4-2
    y=np.cosh(x)+np.cos(x)-g
    plt.scatter(x,y,s=2)
    plt.grid()
    plt.savefig('Graph_g{}.pdf'.format(g))
    plt.clf()
    a=np.random.rand()*5
    b=search(f,a,g)
    if(b==None):
        continue
    elif(b>a):
        c=bisec(f,a,b,g)
        print("Para g={}, f tiene un cero en: [{},{}], exactamente en x={}, teninedo f(x)={}".format(g,b,a,c,f(c,g)))
        continue
    elif(b<a):
        c=bisec(f,a,b,g)
        print("Para g={}, f tiene un cero en: [{},{}], exactamente en x={}, teninedo f(x)={}".format(g,b,a,c,f(c,g)))
        continue

File: Ejercicio_2.1/Solución.cpp
#include <iostream>
#include <cmath>
#include <functional>
#include <cstdlib>

double f(double x, int g);

bool sign(double a, double b);

double search(std::function<double(double,int)> f, double a, int g);

int main(){

  for(int g=2;g<5;g++){
    double a=1/(std::rand()%100)+rand()%10;
    double b=search(f,a,g);
    if(b>a){
      std::cout<<"El cero de f se encuentra en ["<<a<<","<<b<<"] con gamma="<<g<<". f(a)="<<f(a,g)<<" f(b)="<<f(b,g)<<" \n";
    }else if(b<a){
      std::cout<<"El cero de f se encuentra en ["<<b<<","<<a<<"] con gamma="<<g<<". f(a)="<<f(a,g)<<" f(b)="<<f(b,g)<<" \n";
    }
  }
  
  return 0;
}

double f(double x,int g){
  double a=std::cosh(x);
  double b=std::cos(x);
  return a+b-g;
}

bool sign(double a, double b){
  if(a*b<0){
    return true;
  }else if(a*b>0){
    return false;
  }else{
    return false;
  }
}

double search(std::function<double(double,int)> f, double a, int g){
  double ya=f(a,g);
  double bs,bi;
  bs=a;
  bi=a;
  do{
    bs+=0.1;
    bi-=0.1;
    std::cout<<a<<"\t"<<bs<<"\t"<<bi<<"\n";
    if(sign(ya,f(bs,g))){
      return bs;
    }else if(sign(ya,f(bi,g))){
      return bi;
    }
  }while(true);
}
